target/linux/ath79/dts/ar9344_wd_mynet-n750.dts

```diff
@@ -0,0 +1,221 @@
+
+	chosen {
+		bootargs = "console=ttyS0,115200n8";
+	};
+
+	aliases {
+		led-boot = &power;
+		led-failsafe = &power;
+		led-running = &power;
+		led-upgrade = &power;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		wifi {
+			label = "wd:blue:wifi";
+			gpios = <&gpio 11 GPIO_ACTIVE_HIGH>;
+		};
+
+		internet {
+			label = "wd:blue:internet";
+			gpios = <&gpio 12 GPIO_ACTIVE_HIGH>;
+		};
+
+		wps {
+			label = "wd:blue:wps";
+			gpios = <&gpio 13 GPIO_ACTIVE_HIGH>;
+		};
+
+		power: power {
+			label = "wd:blue:power";
+			gpios = <&gpio 14 GPIO_ACTIVE_HIGH>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		reset {
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+
+		wps {
+			linux,code = <KEY_WPS_BUTTON>;
+			gpios = <&gpio 19 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+	};
+
+	gpio-export {
+		compatible = "gpio-export";
+
+		gpio_ext_lna0 {
+			gpio-export,name = "wd:ext:lna0";
+			gpio-export,output = <1>;
+			gpios = <&gpio 15 GPIO_ACTIVE_HIGH>;
+		};
+
+		gpio_ext_lna1 {
+			gpio-export,name = "wd:ext:lna1";
+			gpio-export,output = <1>;
+			gpios = <&gpio 18 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&ref {
+	clock-frequency = <40000000>;
+};
+
+&uart {
+	status = "okay";
+};
+
+&gpio {
+	status = "okay";
+};
+
+&spi {
+	num-cs = <1>;
+
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "bootloader";
+				reg = <0x000000 0x40000>;
+				read-only;
+			};
+
+			partition@40000 {
+				label = "bdcfg";
+				reg = <0x040000 0x10000>;
+				read-only;
+			};
+
+			partition@50000 {
+				label = "devdata";
+				reg = <0x050000 0x10000>;
+				read-only;
+			};
+
+			partition@60000 {
+				label = "devconf";
+				reg = <0x060000 0x10000>;
+				read-only;
+			};
+
+			partition@70000 {
+				compatible = "seama";
+				label = "firmware";
+				reg = <0x070000 0xf80000>;
+			};
+
+			art: partition@ff0000 {
+				label = "art";
+				reg = <0xff0000 0x010000>;
+				read-only;
+			};
+		};
+	};
+};
+
+&usb {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	status = "okay";
+
+	port@1 {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		reg = <1>;
+		#trigger-source-cells = <0>;
+
+		hub_port1: port@1 {
+			reg = <1>;
+			#trigger-source-cells = <0>;
+		};
+
+		hub_port2: port@2 {
+			reg = <2>;
+			#trigger-source-cells = <0>;
+		};
+	};
+};
+
+&usb_phy {
+	status = "okay";
+};
+
+&pcie {
+	status = "okay";
+
+	ath9k: wifi@0,0 {
+		compatible = "pci168c,0033";
+		reg = <0x0000 0 0 0 0>;
+		qca,no-eeprom;
+	};
+};
+
+&wmac {
+	status = "okay";
+	qca,no-eeprom;
+};
+
+&mdio0 {
+	status = "okay";
+	phy-mask = <0>;
+
+	/* PHY 0, MII_BMCR, (BMCR_RESET | BMCR_ANENABLE | BMCR_SPEED1000) */
+	qca,ag71xx-mdio-resetvals = <0 0x00 0x9040>;
```

Comment:
This doesn't look right to me as you're using ar8xxx regs (BMCR) in the ar9344 MDIO block (AG71XX). Probably ar8xxx_phy_check_aneg needs some changes in order to support your device properly as well?

Reply:
See 8b712436f37cea6b25e5a717042ab10b85fab3be and c72313fd9801143029e7381d3df92e49fe45c56b

This register needs to be written to allow the switch to be discovered, before the appropriate ar8xxx driver is in play. Otherwise there is a chicken and egg problem.

Only phy 0 needs to be brought up in my testing, not 0-5 per the previous WD N750 fixup.

Reply:
FYI @olerem has almost [upstreamed the ag71xx driver](https://patchwork.ozlabs.org/patch/1101843/) so I would like to be really careful now and accept only changes to ag71xx which doesn't make his effort harder. I've asked him if he could chime in and provide you with some feedback as well.

From my point of view, kernel folks won't accept this solution as you're fiddling with ar8xxx device inside the ag71xx. Simply ag71xx don't need to care what the underlying ar8xxx switch needs, it should be handled by ar8xxx switch driver itself.

It looks like the call stack is following:

```
phy_probe
  phy_device_reset
     mdio_device_reset
        ar8xxx_phy_soft_reset
           ar8xxx_phy_config_init
              ar8xxx_phy_check_aneg
  ar8xxx_phy_probe
```

So I'm either reading it wrong or I see several places where this could be fixed in ar8xxx.

Reply:
I agree, the code has nothing to do with ag71xx. It is poorly ar8xxx code and should be implemented there.

Reply:
If this problem is just that PHY ID isn't readable before some register operations, you could take a look at my approach for TP-LINK TL-WR941N v7 CN where I added a mdio-device probing for ar8216 switch driver.

Reply:
You can force the phy-id discovery see:

https://www.kernel.org/doc/Documentation/devicetree/bindings/net/phy.txt

```
  If the PHY reports an incorrect ID (or none at all) then the
  "compatible" list may contain an entry with the correct PHY ID in the
  form: "ethernet-phy-idAAAA.BBBB" where
     AAAA - The value of the 16 bit Phy Identifier 1 register as
            4 hex digits. This is the chip vendor OUI bits 3:18
     BBBB - The value of the 16 bit Phy Identifier 2 register as
            4 hex digits. This is the chip vendor OUI bits 19:24,
            followed by 10 bits of a vendor specific ID.
```

Reply:
phy-id did the trick. Thank you all for pointing me in the right direction.

Reply:
I find it quite entertaining, that we've currently (at least?) two ways how to approach this issue. It makes me wonder @981213 if we need both (upstream phy based one and yours mdio one) probably very similar solutions.

I mean, is upstream going to accept the mdio based one if probably the same could be achieved with phy based one, which is already available in the upstream? Maybe it would make sense to convert TL-WR941N v7 CN to that phy based upstream solution and remove the mdio based probing?

Reply:
@ynezz 
Technically a switch isn't a PHY. mdio-device is the upstream framework specifically for these devices that uses mdio bus but isn't a PHY, and that's what's used for almost all upstream DSA drivers. (The only exception is the one for the ancient Marvell 88E6060)

Also, the added mdio-device probing isn't only for WR941N v7. The builtin switch in ar724x/ar933x uses the same PHY ID and chip ID as standalone AR8216 switch while they are actually different, and we need a way to distinguish them (where the compatible string is used for this purpose).